### PR TITLE
[SPARK-6898][SQL] completely support special chars in column names

### PR DESCRIPTION
**Pull request by cloud-fan**
Even if we wrap column names in backticks like ``a#$b.c``,  we still handle the "." inside column name specially. I think it's fragile to use a special char to split name parts, why not put name parts in `UnresolvedAttribute` directly?

### CODE REVIEW

**Comment by cloud-fan** [comment]
This can also resolve https://issues.apache.org/jira/browse/SPARK-5632.
cc @rxin @marmbrus

**Comment by marmbrus** [comment]
ok to test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala [hunk 1]
```diff
@@ -59,7 +64,7 @@ case class UnresolvedAttribute(name: String) extends Attribute with trees.LeafNo
   override def newInstance(): UnresolvedAttribute = this
   override def withNullability(newNullability: Boolean): UnresolvedAttribute = this
   override def withQualifiers(newQualifiers: Seq[String]): UnresolvedAttribute = this
-  override def withName(newName: String): UnresolvedAttribute = UnresolvedAttribute(name)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala]: Origin code ignore the `newName`. Is this intended?

Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala]: No, that seems wrong to me.

**Comment by marmbrus** [comment]
Manually fixed conflicts while merging to master.  Thanks!

**Pull request closed without merging**